### PR TITLE
Grafana: include a built-in backend datasource

### PR DESCRIPTION
**Pull request by ryantxu**
**What this PR does / why we need it**:

In https://github.com/grafana/grafana/pull/38473 and https://github.com/grafana/grafana/pull/37279, we need a UI to let people pick icons and images that are saved in the server.  Currently we have a [static json file](https://play.grafana.org/public/img/icons/unicons/index.json) that lists the icons, but this won't work will for extending the list or letting people upload files.

Rather than add a new API, can we just use a datasource?  On the frontend, we have "-- Grafana --" datasource, but no corresponding backed implementation.

But the datasource needs to be loaded by default with a consistent id/uid for every instance so that frontend components can depend on it.

image

### CODE REVIEW

**Comment by wbrowne** [comment]
Should this just be made a Core plugin @ryantxu? Something similar to [the Test DataSource](https://github.com/grafana/grafana/blob/main/pkg/tsdb/testdatasource/testdata.go)

**Comment by ryantxu** [comment]
> Should this just be made a Core plugin

@wbrowne -- Ideally both test datasource and grafana datasource are enabled/registered and ready out-of-the-box.  The main distinction with this PR is that we accept requests with a fixed id/uid

If there were an interface wrapping the plugin manger, this would be pretty trivial -- but the current form strangely depends on hacking the settings cache :man_shrugging:

File: pkg/services/datasources/cache.go [hunk 1]
```diff
@@ -31,6 +31,14 @@ func (dc *CacheServiceImpl) GetDatasource(
 	user *models.SignedInUser,
 	skipCache bool,
 ) (*models.DataSource, error) {
+	if datasourceID == -1 {
```
Comment (ryantxu) [inline pkg/services/datasources/cache.go:34]: @wbrowne weird hack in an odd place -- any better ideas?

File: pkg/tsdb/service.go [hunk 2]
```diff
@@ -70,8 +77,12 @@ func (s *Service) HandleRequest(ctx context.Context, ds *models.DataSource, quer
 
 		return plugin.DataQuery(ctx, ds, query)
 	}
+	handler := backend.QueryDataHandler(s.BackendPluginManager)
```
Comment (ryantxu) [inline pkg/tsdb/service.go:80]: perhaps there is a way to shove a statically defined plugin into the `BackendPluginManager`?

File: pkg/api/metrics.go [hunk 3]
```diff
@@ -222,37 +231,3 @@ func GenerateSQLTestData(c *models.ReqContext) response.Response {
 
 	return response.JSON(200, &util.DynMap{"message": "OK"})
 }
-
-// GET /api/tsdb/testdata/random-walk
```
Comment (ryantxu) [inline pkg/api/metrics.go:226]: no need for this custom one-off endpoint if we can use the normal route

File: go.mod [hunk 4]
```diff
@@ -116,125 +116,6 @@ require (
 	xorm.io/xorm v0.8.2
 )
 
-require (
```
Comment (ryantxu) [inline go.mod:119]: ?? are these required?  I just ran `go mod tidy`

File: pkg/api/metrics.go [hunk 5]
```diff
@@ -32,31 +30,42 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
 	}
 
 	// Loop to see if we have an expression.
+	prevType := ""
```
Comment (ryantxu) [inline pkg/api/metrics.go:33]: I moved this to its own PR:
https://github.com/grafana/grafana/pull/38949

File: pkg/api/frontendsettings.go [hunk 6]
```diff
@@ -112,11 +113,16 @@ func (hs *HTTPServer) getFSDataSources(c *models.ReqContext, enabledPlugins *plu
 	// the datasource table)
 	for _, ds := range hs.PluginManager.DataSources() {
 		if ds.BuiltIn {
-			dataSources[ds.Name] = map[string]interface{}{
+			info := map[string]interface{}{
 				"type": ds.Type,
 				"name": ds.Name,
 				"meta": hs.PluginManager.GetDataSource(ds.Id),
 			}
+			if ds.Name == grafana.DatasourceName {
```
Comment (ryantxu) [inline pkg/api/frontendsettings.go:121]: @wbrowne -- I put this back so that the front end can treat this like any other data source, and will work fine as we move to looking things up by uid rather than name.  See: https://github.com/grafana/grafana/pull/33817

The more normal this data source can be, the better :)

File: pkg/server/backgroundsvcs/background_services.go [hunk 7]
```diff
@@ -46,7 +47,7 @@ func ProvideBackgroundServiceRegistry(
 	_ *azuremonitor.Service, _ *cloudwatch.CloudWatchService, _ *elasticsearch.Service, _ *graphite.Service,
 	_ *influxdb.Service, _ *loki.Service, _ *opentsdb.Service, _ *prometheus.Service, _ *tempo.Service,
 	_ *testdatasource.TestDataPlugin, _ *plugindashboards.Service, _ *dashboardsnapshots.Service,
-	_ *postgres.Service, _ *mysql.Service, _ *mssql.Service,
+	_ *postgres.Service, _ *mysql.Service, _ *mssql.Service, _ *grafana.Service,
```
Comment (ryantxu) [inline pkg/server/backgroundsvcs/background_services.go:50]: are we at all worried about the name "grafana.Service" as the builtin plugin name?  seems like the scope could be too large.  maybe `grafanads.Service`?

File: pkg/api/frontendsettings.go (hunk 6 above)
Comment (wbrowne) [reply pkg/api/frontendsettings.go:121]: Yeah makes sense 👍 Do weird things start to happen on the frontend if we set uid and id here for all built in data sources?

Comment (ryantxu) [reply pkg/api/frontendsettings.go:121]: In #33817, we set the id/uid equal to the name when not set, so likely a good idea to just do it from the initial response.  For "dashboard" and "mixed" it should never get sent to the backend, but who knows

File: pkg/server/backgroundsvcs/background_services.go (hunk 7 above)
Comment (ryantxu) [reply pkg/server/backgroundsvcs/background_services.go:50]: I am renaming to `grafana_ds.Service` -- I don't see any other `_` but `grafanads` reads to me like "grafa-nads" :grimacing:

File: go.mod [hunk 8]
```diff
@@ -46,7 +46,7 @@ require (
 	github.com/gorilla/websocket v1.4.2
 	github.com/gosimple/slug v1.9.0
 	github.com/grafana/grafana-aws-sdk v0.7.0
-	github.com/grafana/grafana-plugin-sdk-go v0.112.0
+	github.com/grafana/grafana-plugin-sdk-go v0.112.1-0.20210907204529-6d0be6b7893a
```
Comment (ryantxu) [inline go.mod:49]: @wbrowne -- can you review this PR assuming we link it to a new release?  I think it is ready to go.  Thanks for your help!!!

File: pkg/tsdb/grafana_ds/grafana.go [hunk 9]
```diff
@@ -0,0 +1,230 @@
+package grafana_ds
```
Comment (wbrowne) [inline pkg/tsdb/grafana_ds/grafana.go:1]: If not named `grafana` then this should be `grafanads` to align with convention. I thought `grafana` was fine though as it lives under `tsdb`. It goes in line with all the other plugins here 🤷‍♂️

File: pkg/tsdb/grafana_ds/grafana.go [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+func (s *Service) doReadQuery(query backend.DataQuery) backend.DataResponse {
+	q := &listQueryModel{}
+	response := backend.DataResponse{}
+	err := json.Unmarshal(query.JSON, &q)
+	if err != nil {
+		response.Error = err
+		return response
+	}
+
+	if !strings.HasSuffix(q.Path, ".csv") {
```
Comment (wbrowne) [inline pkg/tsdb/grafana_ds/grafana.go:189]: https://pkg.go.dev/path/filepath#Ext is probably more appropriate here

**Review by wbrowne — COMMENTED** [review]
Looks good AFAICT, just some notes. Probably should loop in @marefr since this should look a little clearer now hopefully since the first look. Also @grafana/plugins-platform to check out the frontend

File: pkg/tsdb/grafana_ds/grafana.go (hunk 9 above)
Comment (ryantxu) [reply pkg/tsdb/grafana_ds/grafana.go:1]: nads!   I would like to avoid `grafana.Service` -- I'll change this to `grafanads` and call it a day :)

**Review by wbrowne — APPROVED** [review]
